File: LeetCode/Problem 162.py
# LeetCode: Problem 162
# Jonathan Kosasih

#A peak element is an element that is greater than its neighbors.
#
#Given an input array nums, where nums[i] ≠ nums[i+1], find a peak element and return its index.
#
#The array may contain multiple peaks, in that case return the index to any one of the peaks is fine.
#
#You may imagine that nums[-1] = nums[n] = -∞.
#
#Example 1:
#Input: nums = [1,2,3,1]
#Output: 2
#Explanation: 3 is a peak element and your function should return the index number 2.
#
#Example 2:
#Input: nums = [1,2,1,3,5,6,4]
#Output: 1 or 5 
#Explanation: Your function can return either index number 1 where the peak element is 2, 
#             or index number 5 where the peak element is 6.
#------------------------------------------------------------------
class Solution(object):
    # iterative binary search
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = 0
        r = len(nums)-1
        while l<r:
            p = (l+r) // 2
            if(nums[p]<nums[p+1]):
                l = p+1
            else:
                r = p
        return l
    
sol = Solution()
example = [1,2,1,3,5,6,4]
print(str(sol.findPeakElement(example)))

File: LeetCode/C++/Problem 162.cpp
// LeetCode: Problem 162
// Jonathan Kosasih
/*
A peak element is an element that is greater than its neighbors.

Given an input array nums, where nums[i] ÅÇ nums[i+1], find a peak element and return its index.

The array may contain multiple peaks, in that case return the index to any one of the peaks is fine.

You may imagine that nums[-1] = nums[n] = -Åá.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
	int findPeakElement(const vector<int>& num) {
		int low = 0, high = num.size() - 1;
		while (low < high - 1) {
			int mid = (low + high) / 2;
			if (num[mid] > num[mid - 1] && num[mid] > num[mid + 1])
				return mid;
			else if (num[mid] > num[mid + 1])
				high = mid - 1;
			else
				low = mid + 1;
		}
		return num[low] > num[high] ? low : high;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}